File: C++/654 Maximum Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) { 
        return helper(nums, 0, nums.size() - 1);
    }
    TreeNode* helper(vector<int>& nums, int l, int r) {
        if (l > r) return nullptr;
        int max_index = l;
        for (int i = l; i <= r; i++) {
            if (nums[i] > nums[max_index]) max_index = i;
        }
        TreeNode* root = new TreeNode(nums[max_index]);
        root->left = helper(nums, l, max_index-1);
        root->right = helper(nums, max_index+1, r);
        return root;
    }
};

File: Python/654 Maximum Binary Tree.py
# Maximum Binary Tree
class TreeNode:
    def __init__(self, val = None, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def constructMaximumBinaryTree(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if len(nums) == 0:
            return None
        if len(nums) == 1:
            return TreeNode(nums[0])
        maxnum = nums[0]
        index = 0
        for i, num in enumerate(nums):
            if num > maxnum:
                maxnum = num
                index = i
        root = TreeNode(maxnum)
        root.left = self.constructMaximumBinaryTree(nums[:index])
        root.right = self.constructMaximumBinaryTree(nums[index+1:])
        return root